// vim-language-server/out/handles/completion/identifier.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var config_1 = tslib_1.__importDefault(require("../../server/config"));
var workspaces_1 = require("../../server/workspaces");
var provider_1 = require("./provider");
var util_1 = require("../../common/util");
var patterns_1 = require("../../common/patterns");
function provider(line, uri, position) {
    if (util_1.isSomeMatchPattern(patterns_1.notIdentifierPattern, line)) {
        return [];
    }
    else if (/\b[gbsla]:\w*$/.test(line)) {
        var list = [];
        if (/\bg:\w*$/.test(line)) {
            list = workspaces_1.workspace.getIdentifierItems(uri, position.line)
                .filter(function (item) { return /^g:/.test(item.label); });
        }
        else if (/\bs:\w*$/.test(line)) {
            list = workspaces_1.workspace.getIdentifierItems(uri, position.line)
                .filter(function (item) { return /^s:/.test(item.label); });
        }
        else if (/\bl:\w*$/.test(line)) {
            list = workspaces_1.workspace.getIdentifierItems(uri, position.line)
                .filter(function (item) { return /^l:/.test(item.label); });
        }
        else if (/\ba:\w*$/.test(line)) {
            list = workspaces_1.workspace.getIdentifierItems(uri, position.line)
                .filter(function (item) { return /^a:/.test(item.label); });
        }
        return list.map(function (item) { return (tslib_1.__assign({}, item, { insertText: !/:/.test(config_1.default.iskeyword) ? item.insertText.slice(2) : item.insertText })); });
    }
    else if (/\B:\w*$/.test(line)) {
        return workspaces_1.workspace.getIdentifierItems(uri, position.line)
            .filter(function (item) { return /:/.test(item.label); })
            .map(function (item) {
            var m = line.match(/:[^:]*$/);
            return tslib_1.__assign({}, item, { 
                // delete the `:` symbol
                textEdit: {
                    range: {
                        start: {
                            line: position.line,
                            character: line.length - m[0].length
                        },
                        end: {
                            line: position.line,
                            character: line.length - m[0].length + 1
                        }
                    },
                    newText: item.insertText
                } });
        });
    }
    return workspaces_1.workspace.getIdentifierItems(uri, position.line);
}
provider_1.useProvider(provider);
