// vim-language-server/out/handles/diagnostic.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var vscode_languageserver_1 = require("vscode-languageserver");
var patterns_1 = require("../common/patterns");
var connection_1 = require("../server/connection");
function handleDiagnostic(textDoc, error) {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        var m;
        return tslib_1.__generator(this, function (_a) {
            m = (error || '').match(patterns_1.errorLinePattern);
            if (m) {
                return [2 /*return*/, connection_1.connection.sendDiagnostics({
                        uri: textDoc.uri,
                        diagnostics: [{
                                message: m[1],
                                range: vscode_languageserver_1.Range.create(vscode_languageserver_1.Position.create(parseFloat(m[2]) - 1, parseFloat(m[3]) - 1), vscode_languageserver_1.Position.create(parseFloat(m[2]) - 1, parseFloat(m[3]))),
                                severity: vscode_languageserver_1.DiagnosticSeverity.Error
                            }]
                    })];
            }
            // clear diagnostics
            connection_1.connection.sendDiagnostics({
                uri: textDoc.uri,
                diagnostics: []
            });
            return [2 /*return*/];
        });
    });
}
exports.handleDiagnostic = handleDiagnostic;
